compiler/module_graph.rs

```diff
@@ -0,0 +1,1022 @@
+    };
+
+    Ok(result)
+  }
+
+  /// Update or set the build info for all the roots of a graph.
+  pub fn set_build_info(&mut self, build_info: String) -> Option<String> {
+    self.build_info.insert(self.cache_type.clone(), build_info)
+  }
+
+  /// Set the emitted code for a module in the graph.
+  pub fn set_code(
+    &mut self,
+    specifier: &ModuleSpecifier,
+    code: String,
+    maybe_map: Option<String>,
+  ) -> Result<()> {
+    if let Some(module) = self.modules.get_mut(specifier) {
+      match self.cache_type {
+        CacheType::Cli => {
+          module.code = Some(code);
+          module.map = maybe_map;
+        }
+        CacheType::Bundle => {
+          module.bundle_code = Some(code);
+          module.bundle_map = maybe_map;
+        }
+      }
+      module.is_dirty = true;
+
+      Ok(())
+    } else {
+      Err(MissingSpecifier(specifier.clone()).into())
+    }
+  }
+
+  pub fn transpile(self, options: TranspileOptions) -> Result<CompilerEmit> {
+    if let Some(snapshot_data) = self.maybe_snapshot.clone() {
+      let startup_data = StartupData::Snapshot(Snapshot::Boxed(snapshot_data));
+      let mut compiler = CompilerIsolate::new(startup_data);
+      let graph = Rc::new(RefCell::new(self));
+
+      compiler.transpile(graph, options)
+    } else {
+      Err(MissingSnapshotData.into())
+    }
+  }
+
+  /// Given a set of emitted files, update the modules that are part of the
+  /// graph, marking updated modules as _dirty_ so they will be updated when
+  /// flushed.
+  pub fn update(&mut self, files: &[EmittedFile]) -> Result<()> {
+    for file in files.iter() {
+      if let Some(module_name) = file.maybe_module_name.as_ref() {
+        let specifier = ModuleSpecifier::resolve_url_or_path(module_name)?;
+        if let Some(module) = self.modules.get_mut(&specifier) {
+          let is_map = file.url.ends_with(".map");
+          let data = file.data.clone();
+          match self.cache_type {
+            CacheType::Cli => {
+              if is_map {
+                module.map = Some(data);
+              } else {
+                module.code = Some(data);
+              }
+            }
+            CacheType::Bundle => {
+              if is_map {
+                module.bundle_map = Some(data);
+              } else {
+                module.bundle_code = Some(data);
+              }
+            }
+          };
+          module.is_dirty = true;
+        } else {
+          return Err(MissingSpecifier(specifier).into());
+        }
+      }
+    }
+
+    Ok(())
+  }
+}
+
+/// A structure for building a dependency graph of modules.
+pub struct GraphBuilder {
+  fetched: HashSet<ModuleSpecifier>,
+  graph: Graph,
+  maybe_import_map: Option<Rc<RefCell<ImportMap>>>,
+  pending: FuturesUnordered<FetchFuture>,
+}
+
+impl GraphBuilder {
+  pub fn new(
+    handler: Rc<RefCell<dyn SpecifierHandler>>,
+    maybe_import_map: Option<ImportMap>,
+    maybe_snapshot: Option<Box<[u8]>>,
```

Comment:
Could you explain why does `GraphBuilder` depend on a snapshot? I'm completely lost at this point

Reply:
It isn't the `GraphBuilder` that has a snapshot, it is a `Graph`, maybe... It was a recent API change to contain everything in the graph... so the consumable API is abstracted away fully from the compiler.  You get something like this in the end:

```rust
// If you are doing to do a compile, you provide the snapshot
let mut builder = GraphBuilder::new(handler, maybe_import_map, maybe_snapshot);

// Insert any root specifiers
builder.insert(&specifier).await?;

// Get the graph
let graph = builder.get_graph();

// Compile the graph, which will populate the cache.  No need to worry about the compiler at all
let result = graph.compile(CompileOptions {
  lib: vec!["deno.ns"],
  ..Default::default()
})?;

// Output any ignored options
if let Some(ignored_options) = result.maybe_ignored_options {
  println!("{}", ignored_options);
}
```

But typing that out, I realise I should just move the snapshot to the compile, since that is the only time it is needed.

Reply:
Ok, I have changed the API so that there is `maybe_snapshot` on `Graph::compile()` and `Graph::transpile()`, the only two APIs that might potentially require it, and it keeps that from having to leak anywhere else uneeded.